### ISSUE TITLE
Simplifying the definition of migrations at the type-level

### ISSUE THREAD

**Issue by Raveline** [morphismtech/squeal#54]
One of the key features of Squeal is migration management. Being able to have upgrades and downgrades is great. However, some things remain somehow unclear to me right now.

I'll give some examples taken from my [toy project](https://github.com/Raveline/Melange/blob/master/src/Melange/DB/Schema.hs). It contains a fairly simple model. I defined my table columns in subexpressions, like this one:

```
type QuoteCols =
      '[ "quote_id" ::: 'NoDef :=> 'NotNull 'PGuuid
       , "quote_title" ::: 'NoDef :=> 'Null 'PGtext
       , "content"   ::: 'NoDef :=> 'NotNull 'PGtext
       , "quote_source" ::: 'NoDef :=> 'Null 'PGtext
       ]
```

I want to add a text column named "quote_style". Ideally, I'd like to have a way to start from `QuoteCols` and have a nifty type operator to extend it. I know there is `':` that behaves similarly to `:`, but it would mean I can only add columns at the beginning of the list. It's not very satisfying - it feels odd to have the most recent columns at the beginning of the list, and I'd like to keep my primary key as the first defined field.

Lacking something that would behave similarly to `++`, I rewrote it completely, which is not very DRY:

```
type QuoteColsMig1 =
  '[ "quote_id" ::: 'NoDef :=> 'NotNull 'PGuuid
   , "quote_title" ::: 'NoDef :=> 'Null 'PGtext
   , "content"   ::: 'NoDef :=> 'NotNull 'PGtext
   , "quote_source" ::: 'NoDef :=> 'Null 'PGtext
   , "quote_style" ::: 'NoDef :=> 'Null 'PGtext ]
```

This means I have two definitions for the table. The first one has the initial column definitions, the second the updated one:

```
type QuoteTable =
  "quotes" ::: 'Table (
    '[ "pk_quote" ::: 'PrimaryKey '["quote_id"]
     ] :=> QuoteCols)

type QuoteTableMig1 =
  "quotes" ::: 'Table (
    '[ "pk_quote" ::: 'PrimaryKey '["quote_id"]
     ] :=> QuoteColsMig1)
```

I encounter the same issue as I define the global schema, having to define these two schema, and once again, being not very DRY:

```
type BaseSchema =
    '[ QuoteTable
     , ImageTable
     , ItemsTable
     , BoardTable
     , BoardItemTable
     ]

type Migration1 =
    '[ QuoteTableMig1
     , ImageTableMig1
     , ItemsTable
     , BoardTable
     , BoardItemTable
     ]
```

Though in this instance, I could use `':`.

So, to sum it up: functions that let you modify your schema are easy to use and clear. Using `alterTable` and `addColumn` was simple enough. However, there seems to be no utilities to alter the types that define the schema, which could lead to fairly verbose code and making each migration more expensive.

Am I missing something ? Are there some type operators that ease up defining schema modification at the type level and that I missed ? If not, do you plan to add some ?

I think more examples could be given on this issue. Once again, Squeal's migration system is really a great feature, and I feel it should be showcased more !

**Comment by echatav** [comment]
You can use the DDL type families to accomplish what you're looking for.

```Haskell
>>> :{
type QuoteCols =
      '[ "quote_id" ::: 'NoDef :=> 'NotNull 'PGuuid
       , "quote_title" ::: 'NoDef :=> 'Null 'PGtext
       , "content"   ::: 'NoDef :=> 'NotNull 'PGtext
       , "quote_source" ::: 'NoDef :=> 'Null 'PGtext
       ]
:}
>>> type QuoteColsMig1 = Create "quote_style" ('NoDef :=> 'Null 'PGtext) QuoteCols
>>> :kind! QuoteColsMig1
QuoteColsMig1 :: [(ghc-prim-0.5.1.1:GHC.Types.Symbol,
                   (ColumnConstraint, NullityType))]
= '['("quote_id", 'NoDef :=> 'NotNull 'PGuuid),
    '("quote_title", 'NoDef :=> 'Null 'PGtext),
    '("content", 'NoDef :=> 'NotNull 'PGtext),
    '("quote_source", 'NoDef :=> 'Null 'PGtext),
    "quote_style" ::: '('NoDef, 'Null 'PGtext)]
>>> type QuoteTable = '["pk_quote" ::: 'PrimaryKey '["quote_id"]] :=> QuoteCols
>>> type QuoteTableMig1 = '["pk_quote" ::: 'PrimaryKey '["quote_id"]] :=> QuoteColsMig1
>>> type BaseSchema = '["quotes" ::: 'Table QuoteTable]
>>> type Migration1 = Alter "quotes" ('Table QuoteTableMig1) BaseSchema
>>> :kind! Migration1
Migration1 :: [(ghc-prim-0.5.1.1:GHC.Types.Symbol, SchemumType)]
= '["quotes"
    ::: 'Table
          ('["pk_quote" ::: 'PrimaryKey '["quote_id"]]
           :=> '['("quote_id", 'NoDef :=> 'NotNull 'PGuuid),
                 '("quote_title", 'NoDef :=> 'Null 'PGtext),
                 '("content", 'NoDef :=> 'NotNull 'PGtext),
                 '("quote_source", 'NoDef :=> 'Null 'PGtext),
                 "quote_style" ::: '('NoDef, 'Null 'PGtext)])]
```
>>> :{
let
  migration1up :: Definition BaseSchema Migration1
  migration1up = alterTable #quotes (addColumn #quote_style (nullable text))
:}
>>> :{
let
  migration1down :: Definition Migration1 BaseSchema
  migration1down = alterTable #quotes (dropColumn #quote_style)
:}
>>> import Squeal.PostgreSQL.Migration
>>> :{
let
  migration1 :: Migration IO BaseSchema Migration1
  migration1 = Migration "migration1" (void $ define migration1up) (void $ define migration1down)
:}

**Comment by Raveline** [comment]
Oh, thank you so much ! I'll try to add some example to the type families in the documentation then, the current doc might be a bit confusing for people like me who are not yet very familiar with type families.
Thanks again !
